Optimize BufWriter

tgnottingham opened the pull request.

rust-highfive commented:
r? @m-ou-se

(rust-highfive has picked a reviewer for you, use r? to override)

tgnottingham commented:
This improves BufWriter performance by inlining the hot paths of `write` and `write_all`, not calling expensive `Vec` methods, and avoiding pessimization of common cases.

Optimization was guided by rustc benchmarks. I have a local change that uses `BufWriter` more heavily in rustc. It changes code that wrote to a `Vec`, then wrote that `Vec` to a `File`, to code that writes directly to a `BufWriter<File>`.

Performance was initially completely unworkable.

<details>
<summary><b>Result of switching from baseline to current BufWriter</b></summary>

![bufwriter_unoptimized](https://user-images.githubusercontent.com/3668166/101880785-53876b80-3b48-11eb-831d-e3f3b6f46a0a.png)
</details>

Here is the effect of adding each optimization to the current Bufwriter.

**Optimization 1**: inline hot paths
**Optimization 2**: avoid expensive `Vec` methods
**Optimization 3**: avoid pessimizing common cases

<details>
<summary><b>Result of Optimization 1 versus current BufWriter</b></summary>

![bufwriter_opt1](https://user-images.githubusercontent.com/3668166/101880816-600bc400-3b48-11eb-8f9a-fbdb12807228.png)
</details>

<details>
<summary><b>Result of Optimization 1 + 2 versus Optimization 1</b></summary>

![bufwriter_opt2](https://user-images.githubusercontent.com/3668166/101880853-6a2dc280-3b48-11eb-915f-2159b0ce08ac.png)
</details>

<details>
<summary><b>Result of Optimization 1 + 2 + 3 versus Optimization 1 + 2</b></summary>

![bufwriter_opt3](https://user-images.githubusercontent.com/3668166/101880866-7023a380-3b48-11eb-991b-9f942f9c6b14.png)
</details>

Realize that these are benchmarks of the entirety of compilation, not just `BufWriter` code. So an increase in performance in the benchmarks represents a much larger improvement to `BufWriter` performance, at least according to rustc's use of it.

Obviously optimization 2 is the most impactful, optimization 1 is good, and optimization 3 would be good any other day of the week, but after the other two, it's like, whatever. :)

After all was said and done, the toll of using `BufWriter` heavily in rustc was far more tolerable:

<details>
<summary><b>Result of switching from baseline to optimized BufWriter</b></summary>

![bufwriter_optimized](https://user-images.githubusercontent.com/3668166/101881271-08218d00-3b49-11eb-9193-7da8ecfa4a68.png)
</details>

Still isn't quite where I want it to be though. See #79921 for details.

tgnottingham commented:
I'm aware that #78551 is also making changes in `BufWriter`. I'm hoping that this PR and @Lucretiel's can be reconciled. I'm pretty certain the change to avoid using `Vec` methods can be, but I'm not as sure about the inlining part.

pickfire commented on library/std/src/io/buffered/bufwriter.rs:
```diff
@@ -339,7 +421,13 @@ impl<W: Write> Write for BufWriter<W> {
+                // SAFETY: We checked whether or not the spare capacity was large enough above. If
+                // it was, then we're safe already. If it wasn't, we flushed, making sufficient
+                // room for any input <= the buffer size, which includes this input.
+                unsafe {
+                    bufs.iter().for_each(|b| self.write_to_buffer_unchecked(b));
+                };
```
```suggestion
                bufs.iter().for_each(|b| unsafe { self.write_to_buffer_unchecked(b) });
```
Would this be better?

tgnottingham replied:
I don't have a big preference one way or the other. :shrug:

tgnottingham replied:
Well, I tried it and didn't like the way it looked. Thanks, though. :)

the8472 commented on library/std/src/io/buffered/bufwriter.rs:
```diff
@@ -287,47 +301,115 @@ impl<W: Write> BufWriter<W> {
+    #[inline]
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        // Write to the buffer even if it exactly fills it. Trying to avoid this
+        // adds cost to the overwhelmingly common case where it doesn't happen.
```
That is not quite true. `io::copy` uses a `DEFAULT_BUF_SIZE`-sized buffer internally. `BufWriter::new`  also sets its capacity to `DEFAULT_BUF_SIZE`, which means if you toss a `BufWriter` into `io::copy` then for every single iteration of the copy loop it will try to write exactly as much as needed to fill that buffer, i.e. doubles the `memcpy` traffic.

That said, this can be solved with specialization and #78641 does just that.

tgnottingham replied:
Okay, I can roll that change back. It's the least effective of the optimizations here, and rolling it back may avoid other pathological cases.

tgnottingham replied:
Rolled it back. Thanks @the8472. If you think that this change should be reimplemented once #78641 is done, please feel free to add it.

tgnottingham replied:
I managed to reimplement the optimization in a way that is a win for what is often the common case, while not hurting the case you mentioned or pathological cases much. Latest commit message gives the justification.

pickfire commented on library/std/src/io/buffered/bufwriter.rs:
```diff
@@ -378,39 +384,31 @@ impl<W: Write> BufWriter<W> {
-        // prevents pathological cases for other clients which *always* make writes of this size.
-        // See #72919 and #79930 for more info and a breadcrumb trail.
-        if self.buf.len() + buf.len() <= self.buf.capacity() && buf.len() != self.buf.capacity() {
+        // Use < instead of <= to avoid a needless trip through the buffer in some cases.
```
Which needless trip did you mean?

tgnottingham replied:
The one that we avoid when `self.buf.len() == 0` and `buf.len() == self.buf.capacity()`. I'll update the comments to clarify.

the8472 commented:
Out of curiosity, have you experimented with a smaller hammer by using `#[cold]` or annotating the branches with likely/unlikely instead of `#[inline(never)]`?

Kogia-sima commented:
I have a concern that this optimization is unsound when `buf.len() >= isize::MAX`. It may cause overflow, go into the fast path, and apply copy operation excessing the buffer capacity.

tgnottingham commented:
> Out of curiosity, have you experimented with a smaller hammer by using `#[cold]` or annotating the branches with likely/unlikely instead of `#[inline(never)]`?

I don't think I did--definitely not `#[cold]` anyway. I'll play with them.

Now that you mention it, I'm not quite sure what the right thing to do is here. If I don't separate the functions out, the rustc benchmarks take a pretty big hit. So I know that, for my modified rustc's use of `BufWriter`, it's better if they don't get inlined.

But maybe `#[inline(never)]` is too extreme? Maybe there are some situations where the optimizer can see that it would be better to inline them.

I could just remove the annotations, as currently, the "cold" functions don't get inlined in rustc, even without `#[inline(never)]`. But if they start getting inlined again as a result of some other change (and the lack of annotation), perf could take a hit for workloads where the inlining is harmful.

Anyway, I'm open to suggestions.

tgnottingham commented:
> I have a concern that this optimization is unsound when `buf.len() >= isize::MAX`. It may cause overflow, go into the fast path, and apply copy operation excessing the buffer capacity.

Ah, thank you. Looks like there was a correctness issue already, if I understand correctly, but it didn't lead to unsoundness until my change. I'll fix the problem.

tgnottingham commented:
I think I've addressed the possibility of overflow in `write_vectored`. It was a pre-existing issue, and so a bit outside the original scope of this PR. But I'm hoping that's alright, as fixing it is important for the soundness of the optimizations added by this PR.

rust-log-analyzer commented:
The job `mingw-check` of your PR [failed](https://github.com/rust-lang/rust/runs/1560500597) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://github.com/rust-lang/rust/commit/6b68ae4686fb0be045bd849032c7ca33a0721db4/checks/1560500597/logs), [raw log](https://github.com/rust-lang/rust/commit/6b68ae4686fb0be045bd849032c7ca33a0721db4/checks/1560500597/logs)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[command]/usr/bin/git submodule foreach --recursive git config --local --name-only --get-regexp 'http\.https\:\/\/github\.com\/\.extraheader' && git config --local --unset-all 'http.https://github.com/.extraheader' || :
[command]/usr/bin/git config --local http.https://github.com/.extraheader AUTHORIZATION: basic ***
##[endgroup]
##[group]Fetching the repository
[command]/usr/bin/git -c protocol.version=2 fetch --no-tags --prune --progress --no-recurse-submodules --depth=2 origin +152fb8c4586ccce5c35b1719e700971e851efa01:refs/remotes/pull/79930/merge
---
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/library/std/src/io/buffered/bufwriter.rs at line 436:
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/library/std/src/io/buffered/bufwriter.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
             // `usize`. If the computation overflows, then surely the input cannot fit in our
             // buffer, so we forward to the inner writer's `write_vectored` method to let it
             // handle it appropriately.
-            let saturated_total_len = bufs.iter().fold(0usize, |acc, b|
-                acc.saturating_add(b.len())
-            );
+            let saturated_total_len =
+                bufs.iter().fold(0usize, |acc, b| acc.saturating_add(b.len()));
 
             if saturated_total_len > self.buf.capacity() - self.buf.len() {
                 // Flush if the total length of the input exceeds our buffer's spare capacity.
Build completed unsuccessfully in 0:00:19
== clock drift check ==
  local time: Tue Dec 15 23:34:54 UTC 2020
  network time: Tue, 15 Dec 2020 16:35:21 GMT
```

</details><p></p>

[I'm a bot](https://github.com/rust-lang/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-lang/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

tgnottingham commented:
I've been told by @RalfJung that there's no guarantee yet that the maximum slice size won't increase beyond `isize::MAX` in the future. I'll update this PR to avoid the potential for overflow, even though it's not possible with the current language implementation.

tgnottingham commented:
> > Out of curiosity, have you experimented with a smaller hammer by using `#[cold]` or annotating the branches with likely/unlikely instead of `#[inline(never)]`?
> 
> I don't think I did--definitely not `#[cold]` anyway. I'll play with them.

Btw, recombining the hot/cold code and annotating the branches wasn't helpful. And using `#[inline(never)]` versus `#[cold]` versus no annotation at all didn't change the inlining decisions, at least for the case I'm testing. Still not sure what the best course of action is there.

the8472 commented:
`#[cold]` feels more appropriate semantically, but other places in the standard library also use `#[inline(never)]` as an optimization against code bloat so I guess it doesn't really matter.

mzabaluev commented on library/std/src/io/buffered/bufwriter.rs:
```diff
@@ -287,67 +301,176 @@ impl<W: Write> BufWriter<W> {
         if self.get_ref().is_write_vectored() {
-            let total_len = bufs.iter().map(|b| b.len()).sum::<usize>();
-            if self.buf.len() + total_len > self.buf.capacity() {
+            // We have to handle the possibility that the total length of the buffers overflows
+            // `usize`. If the computation overflows, then surely the input cannot fit in our
```
I don't think it's soundly possible. The total length of buffers in memory must still fit the memory address space, so saturation checks just add unnecessary overhead.

tgnottingham replied:
Thanks, that's a good point. Will be happy to remove this.

tgnottingham replied:
Oh...except the `bufs` argument can technically contain multiple `IoSlice`s that reference the same underlying buffer.

mzabaluev replied:
Good point. It's worth making it bulletproof.

mzabaluev commented:
`#[cold]` and `#[inline]` serve different purposes. `#[cold]` hints the code generator to set branch prediction bias, if supported by the target architecture. I think the `_cold` functions should have both `#[cold]` and `#[inline(never)]`.

mzabaluev commented:
> I think the `_cold` functions should have both `#[cold]` and `#[inline(never)]`.

Or perhaps, just `#[cold]` and let the optimizer decide about inlining. Same applies to `#[inline(always)]`; this attribute is prone to be misused.

tgnottingham commented:
Okay, I've rebased, added a comment about the overflow edge case discussed above, and changed `#[inline(never)]` and `#[inline(always)]` to `#[cold]` and `#[inline]` respectively.

mzabaluev commented on library/std/src/io/buffered/bufwriter.rs:
```diff
@@ -287,67 +301,173 @@ impl<W: Write> BufWriter<W> {
+
+    // SAFETY: Requires `buf.len() <= self.buf.capacity() - self.buf.len()`,
+    // i.e., that input buffer length is less than or equal to spare capacity.
+    #[inline]
+    unsafe fn write_to_buffer_unchecked(&mut self, buf: &[u8]) {
```
Minor nitpick: for private functions, the unadorned `inline` attribute is redundant: the compiler always has access to the function body and is therefore able to make inlining decisions.

tgnottingham replied:
But doesn't this encourage inlining slightly? It adds the `inlinehint` attribute to the LLVM IR anyway.

And if there's a question of whether or not we want to encourage inlining, I think we really do, as the point of the optimization is to inline writes to the internal buffer in hot code. I only separated it out into a function for code reuse. The `#[inline]` is just saying "please don't undo the optimization I'm trying to do". :)

Of course, we could trust the compiler to make the right decision based on context. There are probably cases where `write` and `write_all` are called in cold code where the client doesn't want this function to be inlined. But I'm hesitant to optimize for that case at the risk of undoing the intended optimization. Also, in those cases, I think they actually wouldn't want `write` or `write_all` to be inlined, and would need to do something like wrap them in a function marked `#[inline(never)]` anyway.

bors commented:
:umbrella: The latest upstream changes (presumably #79705) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

tgnottingham commented:
Resolved conflicts.

Dylan-DPC commented:
@m-ou-se this is ready for review

klensy commented on library/std/src/io/buffered/bufwriter.rs:
```diff
@@ -150,7 +155,11 @@ impl<W: Write> BufWriter<W> {
+                    if self.done() {
+                        self.buffer.clear();
+                    } else {
+                        self.buffer.drain(..self.written);
```
Why not `truncate`? from description in https://doc.rust-lang.org/nightly/std/vec/struct.Vec.html#method.truncate it looks like `truncate` should do it better, but better need test it.

klensy replied:
Ah, there we drop elements from left, but truncate drops them from right.

m-ou-se commented on library/std/src/io/buffered/bufwriter.rs:
```diff
@@ -319,6 +333,95 @@ impl<W: Write> BufWriter<W> {
+    // it's likely a sign that the client is using an improperly sized buffer
+    // or their write patterns are somewhat pathological.
+    #[cold]
+    fn write_cold(&mut self, buf: &[u8]) -> io::Result<usize> {
```
Looks like it's assumed this function is only called for `buf.len() >= self.spare_capacity()`. It'd make it easier to read and verify if that assumption was written down. Perhaps as a `debug_assert!()`. (Same below.)

tgnottingham replied:
It's only called in that case, but it still works if that assumption doesn't hold, if I'm not mistaken, so I didn't add a `debug_assert!`. Let me know if you still think I should add it and I'll go ahead.

m-ou-se commented on library/std/src/io/buffered/bufwriter.rs:
```diff
@@ -150,7 +155,11 @@ impl<W: Write> BufWriter<W> {
             fn drop(&mut self) {
                 if self.written > 0 {
-                    self.buffer.drain(..self.written);
+                    if self.done() {
+                        self.buffer.clear();
+                    } else {
```
I'm surprised this makes a difference.

tgnottingham replied:
If it did help, it was negligible. Removed!

m-ou-se commented on library/std/src/io/buffered/bufwriter.rs:
```diff
@@ -319,6 +333,95 @@ impl<W: Write> BufWriter<W> {
+            // exactly. Doing otherwise would mean flushing the buffer, then writing this
+            // input to the inner writer, which in many cases would be a worse strategy.
+
+            // SAFETY: We just called `self.flush_buf()`, so `self.buf.len()` is 0, and
+            // we entered this else block because `buf.len() < self.buf.capacity()`.
```
> We just called `self.flush_buf()`

That's not true when `buf.len() == self.spare_capacity()` but `self.buf.len() > 0`. (Related: Would it be better to check for that case in `write`, before calling `write_cold`, as it doesn't call flush or the underlying write in that case? I assume you tried this too?)

tgnottingham replied:
Thanks! Fixed the comment.

I used to have the check in `write` and `write_all`, but I moved it into the cold functions for the sake of improving performance of what I considered to be the common case. That said, I don't have a strong opinion on this, so I can change it if desired.

m-ou-se reviewed (commented):
This looks good to me. Thank you for working on this, and sorry for the delay.

---

This change makes me wonder whether we should try to extend the interface of `Vec` to support use cases like this where you don't want it to grow (e.g. `Vec::try_push` or `Vec::extend_from_slice_unchecked`, etc.). Or if we should change `Vec::spare_capacity_mut` (which is still unstable) to return some wrapper type that doesn't only give you a `&mut [MaybeUninit]`, but also allows updating `len` somehow. Or maybe a separate `BoundedVec` type would make more sense.

What do you think?

---

A few small comments:

tgnottingham commented:
Rebased on latest master. Will address comments soon.

JohnCSimon commented:
@Rustbot label: +S-waiting-on-review -S-waiting-on-author

tgnottingham commented:
> This change makes me wonder whether we should try to extend the interface of `Vec` to support use cases like this where you don't want it to grow (e.g. `Vec::try_push` or `Vec::extend_from_slice_unchecked`, etc.). Or if we should change `Vec::spare_capacity_mut` (which is still unstable) to return some wrapper type that doesn't only give you a `&mut [MaybeUninit]`, but also allows updating `len` somehow. Or maybe a separate `BoundedVec` type would make more sense.
> 
> What do you think?

In this case, I think it's easy enough to get by with a `Box<[MaybeUninit<u8>]`. In general, I think a non-growing `Vector` might be nice to have, but it would have to be judged against whether or not it's worth adding to the `Vec` API (which is not something I think I'm qualified to evaluate! :)). Seems a bit fringe, but it does come up from time to time.

m-ou-se commented on library/std/src/io/buffered/bufwriter.rs:
```diff
@@ -331,6 +341,103 @@ impl<W: Write> BufWriter<W> {
+    #[inline]
+    unsafe fn write_to_buffer_unchecked(&mut self, buf: &[u8]) {
+        debug_assert!(buf.len() <= self.spare_capacity());
+        let old_len = self.buf.len();
+        let buf_len = buf.len();
+        let src = buf.as_ptr();
+        let dst = self.buf.as_mut_ptr().add(old_len);
+        ptr::copy_nonoverlapping(src, dst, buf_len);
+        self.buf.set_len(old_len + buf_len);
```
(Note that you could also implement this with `self.buf.spare_capacity_mut()` and `MaybeUninit::write_slice`, which are both still unstable.)

m-ou-se commented:
Thanks a lot for doing this!

@bors r+

bors commented:
:pushpin: Commit 01e701828c6010873def5d79191319a6b37a6b32 has been approved by `m-ou-se`

<!-- @bors r=m-ou-se 01e701828c6010873def5d79191319a6b37a6b32 -->
<!-- homu: {"type":"Approved","sha":"01e701828c6010873def5d79191319a6b37a6b32","approver":"m-ou-se"} -->

Mark-Simulacrum commented:
@bors rollup=never - performance PR

bors commented:
:hourglass: Testing commit 01e701828c6010873def5d79191319a6b37a6b32 with merge 676ee14729462585b969bbc52f32c307403f4126...
<!-- homu: {"type":"BuildStarted","head_sha":"01e701828c6010873def5d79191319a6b37a6b32","merge_sha":"676ee14729462585b969bbc52f32c307403f4126"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2522656871)
Approved by: m-ou-se
Pushing 676ee14729462585b969bbc52f32c307403f4126 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"m-ou-se","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2522656871"},"merge_sha":"676ee14729462585b969bbc52f32c307403f4126"} -->
